src/calendarimportmodel.cpp

```diff
@@ -187,66 +173,39 @@ QHash<int, QByteArray> CalendarImportModel::roleNames() const
     roleNames[EndTimeRole] = "endTime";
     roleNames[AllDayRole] = "allDay";
     roleNames[LocationRole] = "location";
+    roleNames[DuplicateRole] = "duplicate";
     roleNames[UidRole] = "uid";
 
     return roleNames;
 }
 
-static bool incidenceLessThan(const KCalendarCore::Incidence::Ptr e1,
-                              const KCalendarCore::Incidence::Ptr e2)
-{
-    if (e1->dtStart() == e2->dtStart()) {
-        int cmp = QString::compare(e1->summary(),
-                                   e2->summary(),
-                                   Qt::CaseInsensitive);
-        if (cmp == 0)
-            return QString::compare(e1->uid(), e2->uid()) < 0;
-        else
-            return cmp < 0;
-    } else {
-        return e1->dtStart() < e2->dtStart();
-    }
-}
-
-void CalendarImportModel::reload()
-{
-    if (!mFileName.isEmpty() || !mIcsRawData.isEmpty()) {
-        bool success = importToMemory(mFileName, mIcsRawData);
-        setError(!success);
-    } else if (!mEventList.isEmpty()) {
-        beginResetModel();
-        mEventList.clear();
-        endResetModel();
-        emit countChanged();
-        setError(false);
-    }
-}
-
 bool CalendarImportModel::importToMemory(const QString &fileName, const QByteArray &icsData)
 {
-    if (!mEventList.isEmpty())
-        mEventList.clear();
-
     bool success = false;
 
-    beginResetModel();
     KCalendarCore::MemoryCalendar::Ptr cal(new KCalendarCore::MemoryCalendar(QTimeZone::systemTimeZone()));
     if (!fileName.isEmpty()) {
         success = CalendarUtils::importFromFile(fileName, cal);
-    } else {
+    } else if (!icsData.isEmpty()) {
         success = CalendarUtils::importFromIcsRawData(icsData, cal);
     }
-    KCalendarCore::Incidence::List incidenceList = cal->incidences();
-    for (int i = 0; i < incidenceList.size(); i++) {
-        KCalendarCore::Incidence::Ptr incidence = incidenceList.at(i);
-        if (incidence->type() == KCalendarCore::IncidenceBase::TypeEvent)
-            mEventList.append(incidence.staticCast<KCalendarCore::Event>());
-    }
-    if (!mEventList.isEmpty())
-        qSort(mEventList.begin(), mEventList.end(), incidenceLessThan);
 
+    beginResetModel();
+    if (!mEventList.isEmpty())
+        mEventList.clear();
+    mEventList = cal->events(KCalendarCore::EventSortStartDate);
+    for (const KCalendarCore::Event::Ptr &event : mEventList) {
```

Comment:
Not really important, but I think this will detach due to ranged-for semantics, because mEventList isn't a const ref.

Reply:
Yes, we need a `qAsConst` here, but it deosn't exist yet in Qt5.6… I left it like that for simplicity, the size of the list is the number of imported events, so it is small. And the list content are QSharedPointers, so the detach should be cheap : no copy will be done.
I've added a comment for future upgrade though.